Initializing git change resolver if dir doesn't contains git folder.

dipak-pawar opened the pull request:
#### Short description of what this resolves:
We are initializing all implementation of `ChangeResolver` using ServiceLoader. Currently if we tries to initialize `GitChangeResolver` it's failing due to `IllegalArgumentException with gitdir is not present` which we haven't catched in either in constrcutor of `GitChangeResolver` or in `isApplicable` method. 

#### Changes proposed in this pull request:

- Catching IllegalArgumentException in constructor of `GitChangeResolver` & `isApplicable` method to avoid unexpected exception while running tests


Fixes #216

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/scm/git/GitChangeResolver.java:
```diff
@@ -62,10 +62,10 @@ public GitChangeResolver(File dir, String previous, String head) {
-        } catch (IOException e) {
-            throw new IllegalArgumentException("Unable to find git repository for path " + dir.getAbsolutePath(), e);
+            this.repoRoot = git.getRepository().getDirectory().getParentFile();
+        } catch (IllegalArgumentException | IOException e) {
```
Isn't there a better way than just catching the exception?
eg: `builder.readEnvironment().findGitDir(dir).getGitDir() != null`

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/scm/git/GitChangeResolver.java:
```diff
@@ -59,23 +59,29 @@ public GitChangeResolver(File projectDir, Scm scm) {
+            }
+            this.repoRoot = git.getRepository().getDirectory().getParentFile();
+        } else {
+            logger.warn("Unable to find git repository for path " + dir.getAbsolutePath());
```
How about throwing an exception when any strategy using scm information is used....?

dipak-pawar replied:
We can throw, but in that case user has to use disable flag to run even `mvn clean` because it's failing while registering extension.

So it's unnecessary burdon for user to use disable flag for mvn clean something like `mvn clean -Dsmart.testing.disable` to make clean runnable

MatousJobanek replied:
Yeah, that's why I said "when any strategy using scm information is used"

dipak-pawar replied:
How about "if user is using any strategy using scm information, then don't add any dependency related to that & warn user to use this strategy you need scm in place." this?

MatousJobanek replied:
well, I think that it should fail, because the ST cannot work properly. At least this is what I would expect as a user.
So, here - at this line - I would warn: `"Unable to find a git repository for the path %s - some strategies won't work properly. Make sure it is a git repository."`
Then, when any strategy tries to get the scm information using method [diff](https://github.com/arquillian/smart-testing/pull/221/files#diff-f1255f2030b502bc7deffb7a6797da92R83) then check if the git is set otherwise throw an exception.

MatousJobanek commented:
btw, is it really WIP as it is in the title of the PR? If yes, then put the label there

dipak-pawar commented:
> btw, is it really WIP as it is in the title of the PR? If yes, then put the label there

It was. Now it's ready

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/scm/git/GitChangeResolver.java:
```diff
@@ -59,23 +59,34 @@ public GitChangeResolver(File projectDir, Scm scm) {
+            }
+            this.repoRoot = git.getRepository().getDirectory().getParentFile();
+        } else {
+            logger.warn(String.format("Unable to find a git repository for the path %s - new, changed, affected strategies won't work properly."
```
I wouldn't mention the concrete names of the strategies - if we change a name or add a new one, we won't probably remember that should be changed in here...

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/scm/git/GitChangeResolver.java:
```diff
@@ -59,23 +59,34 @@ public GitChangeResolver(File projectDir, Scm scm) {
     public Set<Change> diff() {
         final Set<Change> allChanges= new HashSet<>();
+        if (git == null) {
+            throw new IllegalStateException("Git is not initialized. new, changed, affected strategies needs scm to be initialized. "
```
same here with the concrete names. But I understand what you meant - to say which strategies cannot work if the project is not git repository.
There are two ways - return null and then throw the exception from the particular strategy, or provide strategy name as a parameter and incorporate it into the message.

dipak-pawar commented on core/src/test/java/org/arquillian/smart/testing/scm/git/GitChangeResolverTest.java:
```diff
@@ -134,25 +170,29 @@ public void should_fetch_all_modified_files() throws IOException {
         // then
-        assertThat(throwable).isInstanceOf(IllegalArgumentException.class).hasMessageStartingWith("Commit id 'e195e3767591fbd041e041877c541229afaac3c9' is not found in");
+        assertThat(throwable).isInstanceOf(IllegalArgumentException.class)
+            .hasMessageStartingWith("Commit id 'e195e3767591fbd041e041877c541229afaac3c9' is not found in");
```
you mean file or tests. Because I added some tests in this file

MatousJobanek replied:
ok, I meant lines :-)

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/scm/spi/ChangeResolver.java:
```diff
@@ -5,7 +5,7 @@
 public interface ChangeResolver extends AutoCloseable {
 
-    Collection<Change> diff();
+    Collection<Change> diff(String strategy);
```
Here I would create two methods `diff()` and `diff(String strategy)` with the javadoc explaining the usage and why there are two methods - also cases when the exception is thrown. `diff()` will use `null` as the default strategy value

dipak-pawar replied:
I would use `""` as default value for `diff()`

MatousJobanek replied:
Why? Is there any specific reason for it?
From my point of view, there shouldn't be anything set. The string containing something (even empty string) can be already considered as a string containing a content. Which means that if you don't want to set anything (just void) then you should pass the `null` as the best representation of "nothing".

dipak-pawar replied:
just to avoid NPE in future

MatousJobanek replied:
Having the check https://github.com/arquillian/smart-testing/pull/221#discussion_r144232376 if the strategy is null, then you won't get the NPE...
The `GitChangeResolver.diff` is something that can be externally used, so it should properly check if the parameter is null or not, otherwise, you will get unexpected NPE. At least this is what I would expect and how it is usually checked/used. If I check (in case of public classes) whether the string is not provided then as the first thing I check if it is null or not.

MatousJobanek replied:
Anyway, I don't really insist on having `null` as the default value when the `diff()` method is used, but please, put the check (if it is null or not) there. To avoid NPE.

MatousJobanek commented on core/src/main/java/org/arquillian/smart/testing/scm/git/GitChangeResolver.java:
```diff
@@ -59,23 +59,35 @@ public GitChangeResolver(File projectDir, Scm scm) {
-        final Set<Change> allChanges= new HashSet<>();
+    public Set<Change> diff(String strategy) {
+        final Set<Change> allChanges = new HashSet<>();
+        if (!strategy.isEmpty() && git == null) {
```
In case the strategy is `null`, then you get NPE.
In addition, swap the conditions - it makes more sense to check if the git is null as the first condition

dipak-pawar replied:
Here I assumed strategy as `""` not null. that's why I am checking `isEmpty()`

MatousJobanek replied:
see https://github.com/arquillian/smart-testing/pull/221#discussion_r144244237

MatousJobanek commented on core/src/test/java/org/arquillian/smart/testing/scm/git/GitChangeResolverTest.java:
```diff
@@ -25,7 +26,12 @@
 
     @Rule
-    public final TemporaryFolder gitFolder = new TemporaryFolder();
+    public final TemporaryFolder gitFolder = new TemporaryFolder(new File("/tmp"));
```
`/tmp` is the default location of the temporary directory. Then there are subdirectories created by this rule. There is no need to specify the parent directory. In addition, I believe that the location may change based on the system settings

dipak-pawar replied:
If you run tests using surefire it's setting `java.io.Tmp` to module's target dir due to which all temp folders are creating in `submodule/target`. But JGit's `findGitDir(tempDir)` method is checking for all it's parent for `.git` due to which it got `.git` directory which is of smart-testing git & tests fail.

This is not in IntelliJ as it won't set any system property like surefire

MatousJobanek replied:
Ok, I didn't know that. Thx for the explanation. Bloody surefire...
Well, then we should find another way as in some cases this could fail with wrong permissions. Imagine also the case of using Windows - this path doesn't exist there.

MatousJobanek replied:
The solution is using:
`gitFolder = Files.createTempDirectory("junit-").toFile();` 
 that will be created in `/tmp` by default

dipak-pawar replied:
@MatousJobanek This won't work. Because internally `createTempDirectory` with look for system property `java.io.tmpdir` if parent folder not specified.

In our case we are using jboss-parent as  a parent for `smart-testing/pom.xml` which has `java.io.tmpdir` set to `${project.build.directory}` due to which this api will create tmpdir in current porject's build dir.

bartoszmajsak replied:
Is the workaround we discussed offline yesterday working for you?

dipak-pawar replied:
no it's not working. Because We won't get `originial.tmp.dir` as in jboss-parent pom.xml it's defined as `${project.build.directory}`. So it's evaluating ``${project.build.directory}` as original tmp dir

dipak-pawar replied:
I am doing it [this](https://github.com/arquillian/smart-testing/pull/221/files#diff-d0f8b495a0c54a61f1eb2e9f2ed992d4R29) way

MatousJobanek replied:
I've just noticed that when I was trying to fix it locally, I did some additional changes and I was setting it manually to `/tmp` in the code...
But exposing another property via surefire config:
```xml
      <plugin>
        <artifactId>maven-surefire-plugin</artifactId>
        <configuration>
          <systemPropertyVariables>
            <original.java.tmp.dir>${java.io.tmpdir}</original.java.tmp.dir>
          </systemPropertyVariables>
        </configuration>
      </plugin>
```
should work. At least, in my case it returns:
```java
"java.io.tmpdir" -> "/home/mjobanek/arquillian/smart-testing/core/target"
"original.java.tmp.dir" -> "/tmp"
```
So technically, it should be also possible to override the settings from the parent and set the same variable to the original one:
```xml
<java.io.tmpdir>${java.io.tmpdir}</java.io.tmpdir>
```

MatousJobanek commented on core/src/test/java/org/arquillian/smart/testing/scm/git/GitChangeResolverTest.java:
```diff
@@ -59,20 +65,51 @@ public void should_fetch_all_files_from_first_commit_to_given_hash() throws Exce
+    public void should_not_applicable_when_git_repository_is_not_initialized() throws IOException {
+        // given
+        gitFolder.delete();
+        gitFolder.create();
```
to avoid this workaround, move the tests methods to separated class - eg.: `GitChangeResolverWithoutGitTest` where, in the before method, you will just create the temporary directory and will not unpack the repository.

MatousJobanek commented on core/src/test/java/org/arquillian/smart/testing/scm/git/GitChangeResolverTest.java:
```diff
@@ -59,20 +65,51 @@ public void should_fetch_all_files_from_first_commit_to_given_hash() throws Exce
+        thrown.expect(IllegalStateException.class);
+
+        // when
+        gitChangeResolver.diff(CUSTOM);
```
Also, verify the case when the strategy name is not specified: 
`gitChangeResolver.diff()` 
In this case, it should return the empty set.

MatousJobanek commented on mvn-extension/src/main/java/org/arquillian/smart/testing/mvn/ext/SmartTestingMavenConfigurer.java:
```diff
@@ -124,11 +124,13 @@ public void afterSessionEnd(MavenSession session) throws MavenExecutionException
             new JavaSPILoader().all(ChangeResolver.class, ChangeResolver::isApplicable);
-        final Collection<Change> changes = stream(changeResolvers.spliterator(), false).map(ChangeResolver::diff)
+        final Collection<Change> changes = stream(changeResolvers.spliterator(), false)
+            .map(changeResolver -> changeResolver.diff(""))
```
Yeah, don't forget a separated method in the API: `changeResolver.diff()`

bartoszmajsak commented on core/src/main/java/org/arquillian/smart/testing/scm/git/GitChangeResolver.java:
```diff
@@ -78,22 +89,18 @@ private void closeGitIfExists() {
+        return getFileRepositoryBuilder(projectDir).getGitDir() != null;
+    }
+
+    private FileRepositoryBuilder getFileRepositoryBuilder(File currentGitDir) {
```
I would just name it `fileRepositoryBuilder`

MatousJobanek commented:
Please, incorporate my comments:
https://github.com/arquillian/smart-testing/pull/221#discussion_r144545602
https://github.com/arquillian/smart-testing/pull/221#discussion_r144546889

dipak-pawar commented:
@MatousJobanek 
> https://github.com/arquillian/smart-testing/pull/221#discussion_r144545602 (Don't fail when no strategy name is provided)
 
We are using `isApplicable()` method to load all implementations of `ChangeResolver`.  In this PR we changed logic of [isApplicable()](https://github.com/arquillian/smart-testing/pull/221/files#diff-f1255f2030b502bc7deffb7a6797da92R91) such a that it won't fail if you haven't initialized git in your project. Also for this we don't need any strategy.

Now we are calling`Collection<Change> diff(File projectDir, Configuration configuration, String strategy) ` from all implementation of `TestExecutionPlanner`.

Case 1:
If we called it on project which has git initialized, with strategy null, it won't fail, it will give you all required changes as git is initialized.

Case 2:
If we called it on project which has not git initialized, with strategy null, it will fail with `git is not initialized for null strategy`. We are using strategy just for logging msg.



> https://github.com/arquillian/smart-testing/pull/221#discussion_r144546889 (Test for "SmartTestingMavenConfigurer use case")

I can write test for this with strategy `null`, but again it will fail with msg `strategy null needs scm to be initialized` which is same scenario like some random name for which I already added some tests. There is no point to add one more tests for this.

I think we don't need to check this.

MatousJobanek commented:
OK, I missed that part. Thx for explanation @dipak-pawar
